### ISSUE TITLE
Datepicker: need auto open when click on input

### ISSUE THREAD

**Issue by sometime** [angular/material2#8905]
#### Bug, feature request, or proposal:
feature

#### What is the expected behavior?
When i click on input - i want to see calendar without click on icon and input must be focused

#### What is the current behavior?
To view calndar i need click on icon... and input not focused

#### What are the steps to reproduce?
Use datepicker and click on input (not icon) - calendar not opened... when click on icon - calendar opened, but input not focus

**Comment by julianobrasil** [comment]
You can use the `(focus)` and the `(click)` events to programmaticaly open the calendar. Also you can mark the input as `readonly` if you want to prevent users to type into the input field. If I remember right this approach needed a workaround in IE11, but it's not so difficult. I'll open another issue about IE11 behavior.

stackblitz: https://stackblitz.com/edit/open-calendar-on-focus

**Comment by sometime** [comment]
`readonly `is not appropriate: if user want - he can write date on keyboard. And when focus on input, placholder must go to up, like default behavior, but this does not happen.

**Comment by julianobrasil** [comment]
I see... the calendar get the focus as soon as it opens, that's why the label doesn't animate. Well, as a workaround you can get rid of the `readonly` and force the focus back to the input: 

```html
<mat-form-field>
  <input matInput [matDatepicker]="picker"  
         placeholder="Choose a date" 
         (focus)="_openCalendar(picker)" 
         (click)="_openCalendar(picker)" 
         #elementToFocus>
  <mat-datepicker #picker="matDatepicker"></mat-datepicker>
</mat-form-field>
```


```ts
@ViewChild('elementToFocus') _input: ElementRef;

...

_openCalendar(picker: MatDatepicker<Date>) {
    picker.open();
    setTimeout(() => this._input.nativeElement.focus());
}
```

I've updated the [stackblitz](https://stackblitz.com/edit/open-calendar-on-focus) with this.

If you use this component very often, wrap it in a component so you doesn't need to do all this to achieve the UX you're looking for everytime you need a Datepicker.

**Comment by sometime** [comment]
Exactly what is needed! Thank you so much!

**Comment by sometime** [comment]
Found a bug. When i click on input and calendar opened - very good, but if i not select the date and just click on other body empty place (to focus out) - calendar is closed, but cursor again focused on input..  Very not friendly.

**Comment by cephillips117** [comment]
@sometime since you are now triggering the events via the input you would then have to also account for the blur()  I've looked at @julianobrasil stackblitz  example and with some minor changes got it to work: 

`<mat-datepicker #picker="matDatepicker" (closed)="eventCloseHandler($event)"></mat-datepicker>`

`eventCloseHandler(e){
    setTimeout(() => this._input.nativeElement.blur());
  }`

I also don't think it's a bug since what @julianobrasil provided was a work-around in itself.

I hope this helps.

**Comment by sometime** [comment]
No, this not help.

Example from real life: i add two elements to form - input for name and datepicker for birthdate. First, i focus cursor to datepicker and calendar is opened - good... but if i not select the date and mouse click in input with name - input with name not selected... i must click to input name second time.

**Comment by andrewseguin** [comment]
Can you provide a stackblitz or plunker showcasing your use case?

**Comment by cephillips117** [comment]
Ah I see the issue now and you can see it with https://material.angular.io/components/datepicker/examples

When the user opens the calendar whether by using the mat-datepicker-toggle or the implementation above it requires 2 clicks to trigger the focus of another input box.

Steps to reproduce:
1. Choose any Date Picker from the Examples page linked above open the Date Picker popup
2. Click on any other input box, date picker, or button on the page. It require 2 clicks to focus.

**Comment by sometime** [comment]
> Can you provide a stackblitz or plunker showcasing your use case?

yes, https://stackblitz.com/edit/angular-fxsfj2

> It require 2 clicks to focus.

It is very not frendly... if i see user element in interface and click it - i expect that I will be in the context of this element after this click... not after second click

**Comment by julianobrasil** [comment]
@sometime, when you click outside the calendar panel, you're not hitting any element, but the backdrop built by the  `cdk-overlay`. So, the first click just close the panel.

You will need to inspect where the mouse was cliccked and, in the next moment, after the backdrop is dismissed, you can put focus on your next input,

Something like this:

```ts
this._subcription = Observable.fromEvent(document, 'click')
      .subscribe(e => {
        if (e instanceof MouseEvent) {
          const x = e['pageX'];
          const y = e['pageY']
          const element = document.elementFromPoint(x, y);
          setTimeout(() => {
            if (element instanceof HTMLInputElement) {
              (<HTMLInputElement>element).focus();
            }
          });
        }
      });
```

Here is the stackblitz: https://stackblitz.com/edit/open-calendar-on-focus-v2

**Comment by sometime** [comment]
Your stackblitz not working.
I think this method - wrong way and very difficult for simple expected behavior.

The problem in the datepicker core. When calendar is opened - he takes a lot of power about herself. Closing the calendar should not depend on the standard behavior of other user interface elements.
For example: when input (matInput) mouseover - cursor is "text"... but if this form contains datepicker and i open calendar then if input (matInput) mouseover - cursor is "pointer". Why?
I think - the mechanism of opening calendar must not depend of other user interface elements.

**Comment by sometime** [comment]
very good example, how must work datepicker: https://www.primefaces.org/primeng/#/calendar
if i click one datepicker and not select date.. then click another datepicker - first calendar closing and open next calendar... by one click.. very friendly!

**Comment by sometime** [comment]
If i use one material element in form, for example - one datepicker, - this working very good.
But I have a lot of material elements in the one form (two datepickers, two inputs, two autocompletes). And switching between these elements works very unfriendly.

**Comment by julianobrasil** [comment]
@sometime, as the feature you're looking for isn't out-of-the box, if you wrap up the datepicker in a component and hide all the complexity in this new reusable custom component, maybe it makes your life easier after it's ready. But I agree that it should be more simple to do what you're tryng to.

I'll try out to build up something and put in that stackblitz.com and let you know if I manage to do a basic component (without the ControlValueAccessor implementation part). Or maybe using a custom form filed control. If I succeed I post it here (I really don't know when I'll be able to do it).

**Comment by sometime** [comment]
Will be great!

**Comment by julianobrasil** [comment]
@mmalerba, it looks like the backdrop turn the task of wraping up the datepicker in a custom component a hard time (I was trying to to build up a workaround following @sometime's requisite: to be able to close the datepicker and focus other components with a single mouse click). 

I've tried playing around with `document.elementFromPoint(x,y)` to find out which element was clicked to set the focus there _after_ the backdrop was dismissed (by using a `setTimout`), but the MouseEvent just give me awkward coordinates once in a while.

It looks like this kind of thing should be a feature of `MatDatepicker` itself. It takes too much effort to be setup by the developer in this valid use case IMO.

![image](https://media.giphy.com/media/l0CLTliy9u6KQGuMU/giphy.gif)

I don't think it's worth looking at it, but just in case you'd like to see what I've tried to do in order to checkout whether I haven't did anything wrong: https://stackblitz.com/edit/open-calendar-on-focus-v3

**Comment by mmalerba** [comment]
@julianobrasil Yeah currently the datepicker is not set up to easily allow this. I think your implementation is close to working though, there's was just a bug with the way the observables were set up. Here's a fixed version: https://stackblitz.com/edit/open-calendar-on-focus-v3-zkrepy?file=app/custom-datepicker/app-datepicker.component.ts

**Comment by julianobrasil** [comment]
Nice, as always!

**Comment by sometime** [comment]
Thank you!
This works for me.
But i think this must be as option in datepicker... options name for example "auto open on focus", i think it is mut be in datepicker core.

**Comment by mmalerba** [comment]
Yep, will leave this open to track adding the feature at some point. For now thanks @julianobrasil for the workaround :)

**Comment by julianobrasil** [comment]
As little contribution to the workaround, here is with `ControlValueAccessor` implemented: https://stackblitz.com/edit/open-calendar-on-focus-mmalerba-v2

**Comment by Furg** [comment]
@julianobrasil Works really well except using Safari. When you erase an input value manually or using clear button and then select a new date value with the calendar, input placeholder overlaps text value.

Maybe it is related to this issue: #8679

**Comment by julianobrasil** [comment]
@Furg, unfortunately, I have no easy way to test it with safari 😃. The apps I deal with have to work primarily (I think I can say "only") in an environment where people use almost exclusively chrome (updated once a year, or less), so I haven't setup anything to test things against specific browsers other than Chrome, IE11 and Firefox. 

Oh, and concerning this original issue I was just trying to help someone else. This is not a robust solution IMO (it lacks tests and further development). I think the behavior requested here reflects a better way for the things to work. That being said, I have never heard an end user complaining about the need to perform two clicks in this situations. Of course, this is specific to my scenario.

**Comment by julianobrasil** [comment]
@Furg, I've updated all the dependencies in stackblitz sample. Could you check whether it behaves different in safari? https://stackblitz.com/edit/open-calendar-on-focus-mmalerba-v2

**Comment by Furg** [comment]
@julianobrasil The Safari problem still happens with all the dependencies updated, and also with all stackblitz workarounds from this issue.

I know that this is not an official solution, don't worry and thanks for your time. :+1:

**Comment by Satish-Lakhani** [comment]
How about this..?
**(click)="date.open()" (focus)="date.open()"** as shown below.
```
<mat-form-field>
    <input matInput [matDatepicker]="birth_date" formControlName="birth_date" (click)="birth_date.open()" (focus)="birth_date.open()">
    <mat-datepicker-toggle matSuffix [for]="birth_date"></mat-datepicker-toggle>
    <mat-datepicker #birth_date></mat-datepicker>
</mat-form-field>
```

**Comment by julianobrasil** [comment]
@Satish-Lakhani, I used to do it, I took away this approach of some projects of mine because of the focus event. That's great to navigate using tabs, but if you close the calendar, you have to build a workaround to prevent it to reopen as the focus is restored to the original datepicker input. If you switch windows back and forth to the browser after choosing a date, there comes the focus again... 🤣

**Comment by torabian** [comment]
You can use amazing time picker (amazing-time-picker) for this purpose, and it is fully compatible with Angular 2-7 and Material design:

https://github.com/owsolutions/amazing-time-picker

Live demo:
https://owsolutions.github.io/amazing-time-picker

Thanks.

**Comment by DmitryEfimenko** [comment]
I love how people show enthusiasm to come up with a workaround until an official solution is provided! Thank you everyone.
Meanwhile, I've noticed some accessibility issues with the current workarounds:
* In the stackblitz by @julianobrasil (https://stackblitz.com/edit/open-calendar-on-focus-mmalerba-v2) if you click on the input of a datepicker and then you press Tab to go to the next datepicker, it doesn't do that. Instead it just reopens the same datepicker.
* The example by @mmalerba (https://stackblitz.com/edit/open-calendar-on-focus-v3-zkrepy?file=app/custom-datepicker/app-datepicker.component.ts) works a bit better in this regard, but the focus is trapped between the datepickers. I'm not able to tab to the input control (the third control).

**Comment by DmitryEfimenko** [comment]
Also, the calendar does not get updated while you type in the associated input

**Comment by julianobrasil** [comment]
@DmitryEfimenko, usually workarounds aren't finished solutions. They are just start points, from where people can adapt to do whatever they needed when they opened the issue.

**Comment by DmitryEfimenko** [comment]
I certainly realize that. I'm just pointing out edge cases.
By the way, this seems relevant: #13486
